### PR TITLE
App Level Alerting along with centralized alerting 

### PR DESCRIPTION
**Pull request by anuradha-bajpai-google**

### CODE REVIEW

File: quota-notification/src/test/test.iml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bgood) [inline quota-notification/src/test/test.iml:1]: IntelliJ file, let's remove this from the PR.

File: quota-notification/src/main/main.iml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bgood) [inline quota-notification/src/main/main.iml:1]: IntelliJ file, let's remove this from the PR.

File: quota-notification/src/main/java/functions/ConfigureAppAlert.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package functions;
```
Comment (bgood) [inline quota-notification/src/main/java/functions/ConfigureAppAlert.java:1]: Please add license header

File: quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java [hunk 4]
```diff
@@ -0,0 +1,508 @@
+package functions;
```
Comment (bgood) [inline quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java:1]: Please add license header

File: quota-notification/src/main/java/functions/eventpojos/AppAlert.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package functions.eventpojos;
```
Comment (bgood) [inline quota-notification/src/main/java/functions/eventpojos/AppAlert.java:1]: Please add license header

**Review by bgood — CHANGES_REQUESTED** [review]
Quick pass. Just need to remove a couple IntelliJ files and add some license headers.

File: quota-notification/src/test/test.iml (hunk 1 above)
Comment (anuradha-bajpai-google) [reply quota-notification/src/test/test.iml]: Fixed

File: quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java (hunk 4 above)
Comment (anuradha-bajpai-google) [reply quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java:17]: Fixed

File: quota-notification/src/main/java/functions/eventpojos/AppAlert.java (hunk 5 above)
Comment (anuradha-bajpai-google) [reply quota-notification/src/main/java/functions/eventpojos/AppAlert.java:17]: Fixed

**Review by bgood — APPROVED** [review]

File: quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+  /*
+   * API to initialize App Alert table.
+   * This API loads data from csv file in Cloud Storage bucket to BigQuery table
+   * */
+  public static void loadCsvFromGcsToBigQuery(BigQuery bigquery, TableId tableId){
+    // Do not upload data from csv if table already contains data
+    // This batch upload is used only for initializing the data for the first time
+    // for subsequent addition/modification of data, use DML queries in BigQuery table
+    if(!isAppAlertTableEmpty(bigquery) )
+      return;
```
Comment (bgood) [inline quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java:89]: With the return updates to the CSV aren't applied to the table correct?

Would it be possible to truncate the table, reload the csv and channels updated accordingly?  If so what about having the function subscribe to GCS events, so this only happens when the csv is updated?

File: quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+                    + "."
+                    + TABLE
+                    + "` "
+                    +" SET custom_log_metric_id = \""+appAlert.getCustomLogMetricId()+"\", notification_channel_id = \""+appAlert.getNotificationChannelId()+"\", alert_policy_id = \""+appAlert.getAlertPolicyId()+"\" "
+                    +" WHERE app_code = \""+appAlert.getAppCode()+"\""
+            )
+            .setUseLegacySql(false)
+            .build();
+
+    TableResult result = executeBigQueryQuery(bigquery, queryConfig);
```
Comment (bgood) [inline quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java:229]: Unused variable

File: quota-notification/src/main/java/functions/ConfigureAppAlert.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  public static final String APP_ALERT_TABLE = System.getenv("APP_ALERT_TABLE");
+  public static final String CSV_SOURCE_URI = System.getenv("CSV_SOURCE_URI");
+
+  private static final Logger logger = Logger.getLogger(ConfigureAppAlert.class.getName());
+
+  @Override
+  public void service(HttpRequest request, HttpResponse response)
+      throws Exception {
+    response.getWriter().write("App Notification Configuration starting\n");
+    // Initialize client that will be used to send requests.
```
Comment (bgood) [inline quota-notification/src/main/java/functions/ConfigureAppAlert.java:41]: Convert to logging statement?

File: quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
+    if(!isAppAlertTableEmpty(bigquery) )
+      return;
+    // Parse the csv file and verify that
+    // 1. File is available,
+    // 2. There are no records with empty/null project_id, email_id or app_code
+    // 3. There are no duplicates for app_code
+    if(isCSVParseSuccess()){
+      logger.info("CSV parsed successfully! Loading data in BigQuery...");
+    } else {
+      logger.severe("CSV parsing failed. Fix the CSV to proceed.");
```
Comment (bgood) [inline quota-notification/src/main/java/functions/ConfigureAppAlertHelper.java:97]: Does the severe level equate to something that causes the function to a failed run, such that it shows up in the Cloud Functions UI as a failure, so users are more proactively notified?

**Review by bgood — COMMENTED** [review]
Some initial thoughts/questions.

**Pull request merged**